src/memory/buffer_pool.h

```diff
@@ -4,74 +4,168 @@
+     */
+    ErrorCode unpin( const pageId_t pId ) noexcept;
 
     /**
-     * Releases a buffer
-     * @param in bId The buffer to release
-     * @param in tId The id of the transaction 
-     **/
-    void release( const bufferId_t bId, const transactionId_t tId ) noexcept;
+     * Checkpoints the BufferPool to the storage.
+     * 
+     * @return false if the checkpoint was successful, true otherwise.
+     */
+    ErrorCode checkpoint() noexcept;
 
     /**
-     * Pins the buffer
-     * @param in bId The buffer to pin
-     * @param in tId The id of the transaction
-     * @return The handler of the buffer
-     **/
-    Buffer pin( const bufferId_t bId, const transactionId_t tId ) noexcept;
+     * Sets a page as dirty.
+     * 
+     * @param pId pageId_t of the page to be set as dirty.
+     */
+    void setPageDirty( pageId_t pId ) noexcept;
+
+  private:
 
     /**
-     * Checkpoints the BufferPool to the storage
-     **/
-    void checkpoint() noexcept;
+     * Get a pointer to the specified buffer slot.
+     * 
+     * @param bId bufferId_t of the buffer to get a pointer to.
+     * @return Pointer to bId buffer.
+     */
+    char* getBuffer( const bufferId_t bId ) noexcept;
+
+    /**
+     * Returns the bufferId_t of an empty buffer pool slot. In case none is free
+     * Clock Sweep algorithm is performed to evict a page.
+     * 
+     * @param bId bufferId_t of the free pool slot.
+     * @return false if all pages are pinned, true otherwise.
+     */
+    ErrorCode getEmptySlot( bufferId_t& bId ) noexcept;
+
+    struct bufferDescriptor {
+        /**
+         * Number of current references of the page.
+         */
+        uint64_t    m_referenceCount = 0;
+
+        /**
+         * Number of times the stored page has been accessed since it was loaded in the slot.
+         */
+        uint64_t    m_usageCount    = 0;
+
+        /**
+         * Whether the buffer is dirty or not.
+         */
+        bool        m_dirty         = 0;
+
+        /**
+         * pageId_t on disk of the loaded page.
+         */
+        pageId_t    m_pageId        = 0;
+    };
 
     /**
-     * Gets the size of a buffer in KB
-     * @return The size of a buffer in KB
+     * The file storage where this buffer pool will be persisted.
      **/
-    uint64_t bufferSizeKB() const noexcept;
+    FileStorage* p_storage;
 
-  private:
+    /**
+     * Buffer pool.
+     */
+    char* p_pool;
 
     /**
-     * The file storage where this buffer pool will be persisted
-     **/
-    FileStorage* const m_storage;
+     * Buffer descriptors (metadata).
+     */
+    std::vector<bufferDescriptor> m_descriptors;
+
+    /**
+     * Bitset representing whether a Buffer Pool slot is allocated (1) or not (0).
+     */
+    boost::dynamic_bitset<> m_allocationTable;
+
+    /**
+     * Maps pageId_t with its bufferId_t in case it is currently in the Buffer Pool. 
+     */
+    std::map<pageId_t, bufferId_t> m_bufferTable;
```

Comment:
I think the name m_bufferTable is not concise enough. What about renaming it to "m_bufferToPageMap" ?

Reply:
Solved.